step-27: Updated to new interface.

marcfehling opened the pull request:
Successor of #7371. Part of #7515.

This PR simplifies the tutorial code in such a way that it uses the functions in the `SmoothnessEstimator` namespace. Further, it adds possible extensions so users get motivated to try different decision strategies as well as to parallelize this tutorial.

The domain is kept untouched in this rework, but the way how it's generated has changed: instead of re-using the function from `step-14`, now a dimension-independent formulation does the work.

Previously, this tutorial only performed p-refinement. In addition, this PR now also enables p-coarsening as well.

Further, the hp-decision strategy is still the decay of Fourier coefficients one, since the current introduction for this tutorial is very extensive for this particular strategy. I feel it might confuse long-term users to find a different one here (e.g. the one using Legendre coefficients). How do you feel about this?

I still need to work on the results section to update the application output and pictures, which is why I kept this as a draft.

marcfehling commented on examples/step-27/step-27.cc:
```diff
@@ -474,103 +418,70 @@ namespace Step27
+    GridGenerator::subdivided_hyper_cube(cube, 4, -1., 1.);
+
+    std::set<typename Triangulation<dim>::active_cell_iterator> cells_to_remove;
+    for (const auto &cell : cube.active_cell_iterators())
+      for (unsigned int v = 0; v < GeometryInfo<dim>::vertices_per_cell; ++v)
+        for (unsigned int d = 0; d < dim; ++d)
+          if (std::abs(cell->vertex(v)[d]) < .5)
+            cells_to_remove.insert(cell);
```
We might use `std::copy_if` here instead, but I do not know if this would confuse users who are new to C++.

bangerth replied:
I think this is ok.

bangerth commented on examples/step-27/doc/results.dox:
```diff
@@ -209,7 +209,71 @@ include estimating the smoothness not on single cells, but cell assemblies or
+  <li><i>Legendre coefficient decay:</i> This strategy is quite similar
+  to the current one, but uses Legendre series expansion rather than the
+  Fourier one: instead of sinusoids as basis functions, this strategy uses
+  Legendre polynomials. The transition from the Fourier strategy to
```
```suggestion
  Legendre polynomials. Of course, since we approximate the solution using a finite-dimensional
  polynomial on each cell, the expansion of the solution in Legendre polynomials is also finite
  and, consequently, when we talk about the "decay" of this expansion, we can only consider the
  finitely many nonzero coefficients of this expansion, rather than think about it in asymptotic terms.
  But, if we have enough of these coefficients, we can certainly think of the decay of these
  coefficients as characteristic of the decay of the coefficients of the exact solution (which is, in
  general, not polynomial and so will have an infinite Legendre expansion), and considering the
  coefficients we have should reveal something about the properties of the exact solution.
  
  The transition from the Fourier strategy to
```

bangerth commented on examples/step-27/doc/results.dox:
```diff
@@ -209,7 +209,71 @@ include estimating the smoothness not on single cells, but cell assemblies or
+  <li><i>Refinement history:</i> The last strategy is quite different
+  from the other two. In theory, we know how the error will converge
+  after changing the discretization of the function space. With
+  h-refinement, the solution converges algebraically as already pointed
```
```suggestion
  $h$-refinement: the solution converges algebraically as already pointed
```
Maybe say what "algebraically" means?

bangerth commented on examples/step-27/doc/results.dox:
```diff
@@ -209,7 +209,71 @@ include estimating the smoothness not on single cells, but cell assemblies or
+  from the other two. In theory, we know how the error will converge
+  after changing the discretization of the function space. With
+  h-refinement, the solution converges algebraically as already pointed
+  out in step-7. It the solution is sufficiently smooth, though, we
```
```suggestion
  out in step-7. If the solution is sufficiently smooth, though, we
```

bangerth commented on examples/step-27/doc/results.dox:
```diff
@@ -209,7 +209,71 @@ include estimating the smoothness not on single cells, but cell assemblies or
+  expect that the solution will converge exponentially with increasing
+  polynomial degree of the finite element. We can compare a proper
+  prediction of the error with the actual error in the following step to
+  see verify if our choice of adaptation type was justified. The transition
```
```suggestion
  see if our choice of adaptation type was justified. The transition
```

bangerth commented on examples/step-27/doc/results.dox:
```diff
@@ -209,7 +209,71 @@ include estimating the smoothness not on single cells, but cell assemblies or
+  prediction of the error with the actual error in the following step to
+  see verify if our choice of adaptation type was justified. The transition
+  to this strategy is a bit more complicated. For this, we need an
+  initialization step with pure h- or p-refinement and we need to transfer
```
```suggestion
  initialization step with pure $h$- or $p$-refinement and we need to transfer
```

bangerth commented on examples/step-27/doc/results.dox:
```diff
@@ -209,7 +209,71 @@ include estimating the smoothness not on single cells, but cell assemblies or
+  of the hp::Refinement::predict_error() function describes not only the
+  theoretical details of this approach, but also presents a blueprint on
+  how to implement this strategy in your code. Note that this strategy
+  is only suited for time-independent problems or to create an initial
```
I don't think I understand this comment. I *think* that what you're trying to say is that one can't predict the smoothness of the solution at the next time step, but if one wanted to build the mesh for just one time step (e.g., start each step with a coarse mesh, then keep refining until we're happy with the solution, and only then moving on to the next time step), then this approach here could be used as well.

I think that your comment is not actually specific to this strategy but all the others as well.

marcfehling replied:
You are right, this short comment does not give the issue justice that I wanted to address. I've extended the documentation.

bangerth commented on examples/step-27/doc/results.dox:
```diff
@@ -209,7 +209,71 @@ include estimating the smoothness not on single cells, but cell assemblies or
+Try implementing one of these strategies into this tutorial and observe the
+subtle changes to the results. You will notice that all strategies are
+capable of identifying the singularities near the reentrant corners and
+will perform h-refinement in these regions, while preferring p-refinement
```
```suggestion
will perform $h$-refinement in these regions, while preferring $p$-refinement
```

bangerth commented on examples/step-27/doc/results.dox:
```diff
@@ -209,7 +209,71 @@ include estimating the smoothness not on single cells, but cell assemblies or
+subtle changes to the results. You will notice that all strategies are
+capable of identifying the singularities near the reentrant corners and
+will perform h-refinement in these regions, while preferring p-refinement
+in the bulk domain.
```
I think it would be fair to cite your PhD thesis somewhere here. You *did* implement all of these things, and it makes sense to point the reader to a place where they can find the kind of comparison you're suggesting here.

marcfehling replied:
Yes, we could do this or point to the parallel hp-paper (if we decide to describe them there, too).

bangerth replied:
For now, just reference your thesis. We can always add other references at a later time.

bangerth reviewed (commented):
Just the documentation part. I'll get back to the `.cc` at some later time, but now I have to go meet you (@marcfehling) for a bike ride :-)

bangerth commented:
/rebuild

bangerth commented on examples/step-27/step-27.cc:
```diff
@@ -49,17 +50,19 @@
-// the introduction of this program. The last one provides Fourier
-// transformation class on the unit cell.
+// the introduction of this program. The next one provides the functionality
+// for automatic hp-adaptation, for which we will use the estimation algorithms
```
```suggestion
// for automatic $hp$-adaptation, for which we will use the estimation algorithms
```

bangerth commented on examples/step-27/step-27.cc:
```diff
@@ -474,103 +418,70 @@ namespace Step27
+      // the threshold, and (iii) we still have a finite element with a
+      // polynomial degree higher than the current one in the finite element
+      // collection, we will assign a future FE index that corresponds to a by
+      // one higher polynomial degree. The following function is capable of
```
...that corresponds to a polynomial with degree one higher than it currently is

bangerth commented on examples/step-27/step-27.cc:
```diff
@@ -474,103 +418,70 @@ namespace Step27
+      // threshold as the mean value between minimal and maximal smoothness
+      // indicators on cells flagged for refinement. This is achieved by setting
+      // the corresponding fraction parameter to a value of 0.5. We deal the
+      // same way with cells that are going to be coarsened and decrease their
```
-> In the same way, we deal with cells that are going...

bangerth commented on examples/step-27/step-27.cc:
```diff
@@ -474,103 +418,70 @@ namespace Step27
+        dof_handler, smoothness_indicators, 0.5, 0.5);
+
+      // The above function only determines whether the polynomial degree will
+      // change via future FE indices, but does not manipulate the h-refinement
```
```suggestion
      // change via future FE indices, but does not manipulate the $h$-refinement
```

bangerth commented on examples/step-27/step-27.cc:
```diff
@@ -474,103 +418,70 @@ namespace Step27
+      // The above function only determines whether the polynomial degree will
+      // change via future FE indices, but does not manipulate the h-refinement
+      // flags. So for cells that are flagged for both refinement categories, we
+      // prefer p- over h-refinement. The following function call ensures that
```
```suggestion
      // prefer $p$- over $h$-refinement. The following function call ensures that
```

bangerth commented on examples/step-27/step-27.cc:
```diff
@@ -474,103 +418,70 @@ namespace Step27
+      // change via future FE indices, but does not manipulate the h-refinement
+      // flags. So for cells that are flagged for both refinement categories, we
+      // prefer p- over h-refinement. The following function call ensures that
+      // only either p- or h-refinement is imposed, and not both at once.
```
```suggestion
      // only one of $p$- or $h$-refinement is imposed, and not both at once.
```

bangerth commented on examples/step-27/step-27.cc:
```diff
@@ -474,103 +418,70 @@ namespace Step27
+  // this domain which is dimension independent.
+  //
+  // We first create a hypercube triangulation with enough cells so that it
+  // already holds our desired domain. We then continue to remove those cells in
```
```suggestion
  // already holds our desired domain $[-1,1]^d$, subdivided into $4^d$ cells. We then remove those cells in
```

bangerth commented on examples/step-27/step-27.cc:
```diff
@@ -474,103 +418,70 @@ namespace Step27
+  //
+  // We first create a hypercube triangulation with enough cells so that it
+  // already holds our desired domain. We then continue to remove those cells in
+  // the center of the domain by identifying them with the physical coordinates
```
```suggestion
  // the center of the domain by testing the coordinate values
```

bangerth reviewed (commented):
Nice job removing all of this code!

marcfehling commented:
Addressed your comments. As we keep the Fourier strategy, I will generate updated output and figures now.

Who do I need to talk to in order to upload updated result images? :)

bangerth commented:
You'd talk to me :-)

masterleinad commented:
Is this still WIP?

marcfehling commented:
> Is this still WIP?

Unfortunately yes :(

While preparing updated results, I realized that there are issues with the Fourier strategy as the generated mesh now looks far different than the old one. My best guess that the configuration of the `FESeries::Fouier` object is not sufficient. I'll figure out the cause of this...

bangerth commented:
Let us know once you've got this ready. I'd also be happy to just merge this as-is, and then take the remaining adjustments as a follow-up.

marcfehling commented:
> Let us know once you've got this ready. I'd also be happy to just merge this as-is, and then take the remaining adjustments as a follow-up.

Fair enough! Let's address details about the hp-parameters and the output in a separate PR then. I'll release this PR from its WIP state.
